File: workspace/nodejs/src/services/user.service.ts
/**
 * User Service
 * Business logic for user management
 */

import { User } from '@prisma/client';
import { prisma } from '../utils/database';
import { hashPassword } from '../utils/security';
import { cacheManager } from '../utils/cache';
import { AppError } from '../utils/errors';
import { logger } from '../utils/logger';
import { CreateUserDTO, UpdateUserDTO, UserQueryDTO } from '../dto/user.dto';

export class UserService {
  /**
   * Create new user
   */
  async createUser(userData: CreateUserDTO): Promise<User> {
    // Check if email exists
    const existingEmail = await this.getUserByEmail(userData.email);
    if (existingEmail) {
      throw new AppError(`Email '${userData.email}' is already registered`, 400);
    }

    // Check if username exists
    const existingUsername = await this.getUserByUsername(userData.username);
    if (existingUsername) {
      throw new AppError(`Username '${userData.username}' is already taken`, 400);
    }

    // Hash password
    const hashedPassword = await hashPassword(userData.password);

    // Create user
    const user = await prisma.user.create({
      data: {
        email: userData.email.toLowerCase(),
        username: userData.username.toLowerCase(),
        hashedPassword,
        fullName: userData.fullName,
        isActive: true,
        isVerified: false,
        isSuperuser: false,
      },
    });

    logger.info('User created', { userId: user.id, username: user.username });

    return user;
  }

  /**
   * Get user by ID
   */
  async getUserById(userId: number): Promise<User | null> {
    // Try cache first
    const cacheKey = `user:${userId}`;
    const cached = await cacheManager.get<User>(cacheKey);

    if (cached) {
      logger.debug('User found in cache', { userId });
      return cached;
    }

    // Query database
    const user = await prisma.user.findFirst({
      where: {
        id: userId,
        deletedAt: null,
      },
    });

    // Cache result
    if (user) {
      await cacheManager.set(cacheKey, user, 300); // 5 minutes
    }

    return user;
  }

  /**
   * Get user by email
   */
  async getUserByEmail(email: string): Promise<User | null> {
    return prisma.user.findFirst({
      where: {
        email: email.toLowerCase(),
        deletedAt: null,
      },
    });
  }

  /**
   * Get user by username
   */
  async getUserByUsername(username: string): Promise<User | null> {
    return prisma.user.findFirst({
      where: {
        username: username.toLowerCase(),
        deletedAt: null,
      },
    });
  }

  /**
   * Get paginated list of users
   */
  async getUsers(query: UserQueryDTO): Promise<{ users: User[]; total: number }> {
    const where: any = {
      deletedAt: null,
    };

    if (query.isActive !== undefined) {
      where.isActive = query.isActive;
    }

    // Get total count
    const total = await prisma.user.count({ where });

    // Get users with pagination
    const users = await prisma.user.findMany({
      where,
      skip: query.skip,
      take: query.take,
      orderBy: {
        createdAt: 'desc',
      },
    });

    return { users, total };
  }

  /**
   * Update user
   */
  async updateUser(userId: number, updateData: UpdateUserDTO): Promise<User | null> {
    const user = await this.getUserById(userId);
    if (!user) {
      return null;
    }

    // Check email uniqueness
    if (updateData.email && updateData.email !== user.email) {
      const existingEmail = await this.getUserByEmail(updateData.email);
      if (existingEmail) {
        throw new AppError(`Email '${updateData.email}' is already registered`, 400);
      }
    }

    // Check username uniqueness
    if (updateData.username && updateData.username !== user.username) {
      const existingUsername = await this.getUserByUsername(updateData.username);
      if (existingUsername) {
        throw new AppError(`Username '${updateData.username}' is already taken`, 400);
      }
    }

    // Prepare update data
    const data: any = {};

    if (updateData.email) data.email = updateData.email.toLowerCase();
    if (updateData.username) data.username = updateData.username.toLowerCase();
    if (updateData.fullName !== undefined) data.fullName = updateData.fullName;
    if (updateData.password) data.hashedPassword = await hashPassword(updateData.password);
    if (updateData.isActive !== undefined) data.isActive = updateData.isActive;

    // Update user
    const updatedUser = await prisma.user.update({
      where: { id: userId },
      data,
    });

    // Invalidate cache
    await cacheManager.delete(`user:${userId}`);

    logger.info('User updated', { userId, username: updatedUser.username });

    return updatedUser;
  }

  /**
   * Delete user (soft delete)
   */
  async deleteUser(userId: number): Promise<boolean> {
    const user = await this.getUserById(userId);
    if (!user) {
      return false;
    }

    // Soft delete
    await prisma.user.update({
      where: { id: userId },
      data: {
        deletedAt: new Date(),
        isActive: false,
      },
    });

    // Invalidate cache
    await cacheManager.delete(`user:${userId}`);

    logger.info('User deleted', { userId, username: user.username });

    return true;
  }

  /**
   * Verify user email
   */
  async verifyUserEmail(userId: number): Promise<User | null> {
    const user = await this.getUserById(userId);
    if (!user) {
      return null;
    }

    const updatedUser = await prisma.user.update({
      where: { id: userId },
      data: {
        isVerified: true,
        emailVerifiedAt: new Date(),
      },
    });

    // Invalidate cache
    await cacheManager.delete(`user:${userId}`);

    logger.info('User email verified', { userId, username: updatedUser.username });

    return updatedUser;
  }

  /**
   * Update last login timestamp
   */
  async updateLastLogin(userId: number): Promise<void> {
    await prisma.user.update({
      where: { id: userId },
      data: {
        lastLoginAt: new Date(),
      },
    });

    // Invalidate cache
    await cacheManager.delete(`user:${userId}`);
  }
}

File: workspace/python/app/services/user_service.py
"""
User Service
Business logic for user management
"""

from typing import Optional, List, Tuple
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.user import User
from app.schemas.user import UserCreate, UserUpdate
from app.utils.security import get_password_hash
from app.utils.logger import get_logger
from app.utils.cache import cache_manager


logger = get_logger(__name__)


class UserService:
    """Service class for user operations"""

    def __init__(self, db: AsyncSession):
        self.db = db

    async def create_user(self, user_data: UserCreate) -> User:
        """
        Create a new user

        Args:
            user_data: User creation data

        Returns:
            Created user object

        Raises:
            ValueError: If email or username already exists
        """
        # Check if email exists
        existing_user = await self.get_user_by_email(user_data.email)
        if existing_user:
            raise ValueError(f"Email '{user_data.email}' is already registered")

        # Check if username exists
        existing_user = await self.get_user_by_username(user_data.username)
        if existing_user:
            raise ValueError(f"Username '{user_data.username}' is already taken")

        # Hash password
        hashed_password = get_password_hash(user_data.password)

        # Create user
        user = User(
            email=user_data.email,
            username=user_data.username,
            full_name=user_data.full_name,
            hashed_password=hashed_password,
            is_active=True,
            is_verified=False,
            is_superuser=False,
        )

        self.db.add(user)
        await self.db.commit()
        await self.db.refresh(user)

        logger.info(f"User created: {user.username} (ID: {user.id})")

        return user

    async def get_user_by_id(self, user_id: int) -> Optional[User]:
        """
        Get user by ID

        Args:
            user_id: User ID

        Returns:
            User object or None
        """
        # Try cache first
        cache_key = f"user:{user_id}"
        cached_user = await cache_manager.get(cache_key)

        if cached_user:
            logger.debug(f"User {user_id} found in cache")
            return cached_user

        # Query database
        result = await self.db.execute(
            select(User).where(User.id == user_id, User.deleted_at.is_(None))
        )
        user = result.scalar_one_or_none()

        # Cache result
        if user:
            await cache_manager.set(cache_key, user, ttl=300)  # 5 minutes

        return user

    async def get_user_by_email(self, email: str) -> Optional[User]:
        """
        Get user by email

        Args:
            email: User email

        Returns:
            User object or None
        """
        result = await self.db.execute(
            select(User).where(User.email == email.lower(), User.deleted_at.is_(None))
        )
        return result.scalar_one_or_none()

    async def get_user_by_username(self, username: str) -> Optional[User]:
        """
        Get user by username

        Args:
            username: Username

        Returns:
            User object or None
        """
        result = await self.db.execute(
            select(User).where(User.username == username.lower(), User.deleted_at.is_(None))
        )
        return result.scalar_one_or_none()

    async def get_users(
        self, skip: int = 0, limit: int = 100, is_active: Optional[bool] = None
    ) -> Tuple[List[User], int]:
        """
        Get paginated list of users

        Args:
            skip: Number of records to skip
            limit: Maximum number of records to return
            is_active: Filter by active status

        Returns:
            Tuple of (users list, total count)
        """
        # Build query
        query = select(User).where(User.deleted_at.is_(None))

        if is_active is not None:
            query = query.where(User.is_active == is_active)

        # Get total count
        count_query = select(func.count()).select_from(User).where(User.deleted_at.is_(None))
        if is_active is not None:
            count_query = count_query.where(User.is_active == is_active)

        total_result = await self.db.execute(count_query)
        total = total_result.scalar()

        # Get users with pagination
        query = query.offset(skip).limit(limit).order_by(User.created_at.desc())
        result = await self.db.execute(query)
        users = result.scalars().all()

        return list(users), total

    async def update_user(self, user_id: int, user_data: UserUpdate) -> Optional[User]:
        """
        Update user information

        Args:
            user_id: User ID
            user_data: Update data

        Returns:
            Updated user object or None

        Raises:
            ValueError: If email or username already exists
        """
        user = await self.get_user_by_id(user_id)
        if not user:
            return None

        # Check email uniqueness
        if user_data.email and user_data.email != user.email:
            existing_user = await self.get_user_by_email(user_data.email)
            if existing_user:
                raise ValueError(f"Email '{user_data.email}' is already registered")
            user.email = user_data.email

        # Check username uniqueness
        if user_data.username and user_data.username != user.username:
            existing_user = await self.get_user_by_username(user_data.username)
            if existing_user:
                raise ValueError(f"Username '{user_data.username}' is already taken")
            user.username = user_data.username

        # Update other fields
        if user_data.full_name is not None:
            user.full_name = user_data.full_name

        if user_data.password:
            user.hashed_password = get_password_hash(user_data.password)

        if user_data.is_active is not None:
            user.is_active = user_data.is_active

        await self.db.commit()
        await self.db.refresh(user)

        # Invalidate cache
        cache_key = f"user:{user_id}"
        await cache_manager.delete(cache_key)

        logger.info(f"User updated: {user.username} (ID: {user.id})")

        return user

    async def delete_user(self, user_id: int) -> bool:
        """
        Soft delete user

        Args:
            user_id: User ID

        Returns:
            True if deleted, False if not found
        """
        user = await self.get_user_by_id(user_id)
        if not user:
            return False

        # Soft delete
        user.soft_delete()
        user.is_active = False

        await self.db.commit()

        # Invalidate cache
        cache_key = f"user:{user_id}"
        await cache_manager.delete(cache_key)

        logger.info(f"User deleted: {user.username} (ID: {user.id})")

        return True

    async def verify_user_email(self, user_id: int) -> Optional[User]:
        """
        Verify user email

        Args:
            user_id: User ID

        Returns:
            Updated user or None
        """
        user = await self.get_user_by_id(user_id)
        if not user:
            return None

        user.verify_email()
        await self.db.commit()
        await self.db.refresh(user)

        # Invalidate cache
        cache_key = f"user:{user_id}"
        await cache_manager.delete(cache_key)

        logger.info(f"User email verified: {user.username} (ID: {user.id})")

        return user
